### ISSUE TITLE
error message for casting with "type" defined types needs help

### ISSUE THREAD

**Issue by apinski-cavium** [p4lang/p4c#3221]
Take:
```
typedef bit<1> b1;
typedef bit<2> b2;
type b1 t1;
type b2 t2;

t1 func(b2 a)
{
  return (t1)a;
}
```
Currently we get:
```
casttype.p4(8): [--Werror=type-error] error: cast
  return (t1)a;
         ^^^^^
  ---- Actual error:
  Cannot unify type 'bit<2>' with type 'bit<1>'
  ---- Originating from:
  Cannot cast from 'bit<2>' to 'bit<1>'
```
But casting betweeen bit<2> and bit<1> is valid (according to 8.9.1) but what is not valid is cast between bit<2> and t1 (which is a type defined via type of bit<1>).

Either we make type act similar to enum with an underlying type (which is a specification change) or the error message needs to be improved.

**Comment by jnfoster** [comment]
To be clear, this program is definitely allowed (and is accepted by `p4c`).
```
typedef bit<1> b1;
typedef bit<2> b2;
type b1 t1;
type b2 t2;

t1 func(b2 a)
{
  return (t1) ((b1) a);
}
```
One possible solution would be to say that the set of allowed explicit casts is transitively closed: if you can cast from T1 to T2, and from T2 to T3, then you can also cast from T1 to T3 (without having to mention T2).

**Comment by mbudiu-vmw** [comment]
Unfortunately that's not true for bit/int values and we even have an example. That's because you can't guess the intermediate types and the result is different depending on the guess.

**Comment by jnfoster** [comment]
Of course... Val Tannen had much to say about this kind of (in)coherence in his papers from the late 1980s.

Let me rephrase: making the set of casts be transitively closed (but explicitly picking the semantics for T1 -> T3 when there are several T2s in the middle) might be useful as a design principle.

Here's are some minimal examples showing the essence of the issue.

This code doesn't type check, with the same error message as above.
```
type bit<1> t1;

t1 func(bit<2> a)
{
  return (t1)a;
}
```

This code, with an extra cast from `bit<2>` to `bit<1>`, does type check.
```p4
type bit<1> t1;

t1 func(bit<2> a)
{
  return (t1) (bit<1>) a;
}
```

And this code, with an integer literal, also type checks (you might wonder about it since there is an implicit conversion to `bit<1>` happening for the integer literal).
```
type bit<1> t1;

t1 func(bit<2> a)
{
  return (t1) 1;
}
```
